### ISSUE TITLE
Inconsistent osd_memory_target defined in OSD vs MON

### ISSUE THREAD

**Issue by kbmanseau** [rook/rook#9110]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
Expected behavior appears to be that if unset or if pod limits are not specified, osd_memory_target will be set to 4GB by default. This appears to be true when checking from within the ceph-tools pod (believe it's polling from the mon)
`[root@rook-ceph-tools-6967fc698d-4w2wk /]# ceph config get osd.1 osd_memory_target
4294967296`

However when checking with the ceph tell command (polling the daemon directly?), the value is reported differently
`[root@rook-ceph-tools-6967fc698d-4w2wk /]# ceph tell osd.1 config show | grep osd_memory_target
    "osd_memory_target": "432630831513",`
Interestingly though this value is not the max value of memory on the node, perhaps this was the amount of free memory polled upon the start of the OSD? Similar results are seen on each OSD.

When performing a dump_mempools, I can see the bluestore_cache_data and bluestore_cache_other both consuming ~25GB of memory. All of our OSD vary between a current usage of 50GB-150GB of memory usage as seen through our monitoring and confirmed by checking consumed memory on each node.

We observe a ceph crash on a per-node basis whenever the memory usage gets too high from the OSDs. I assume it's because there are no limits set to perform OOM killing. 

The workaround/solution in my eyes is to set a limit as describe in the config [here](https://github.com/rook/rook/blob/41ef578859f23b87ff9002105000a864dd9dbc2c/cluster/examples/kubernetes/ceph/cluster.yaml#L182) which is what we will be doing next. Just wanted to capture my findings and gather thoughts before doing so

**Expected behavior:**
Default should result in consistent settings of the osd_memory_target value across OSDs. 

**Environment**:
* 3x centos 8 nodes (kernel 5.9.6-1)
    *  503Gi of Memory per node
* k8s 1.19.3
* Rook version: 1.4.2
* Storage backend version: ceph 15.2.4
* Kubernetes cluster type: Self-hosted baremetal
* Storage backend status: HEALTH_OK

**Comment by travisn** [comment]
@kbmanseau The value actually used by the OSD may not be accurate as reported from the toolbox. If you connect to an OSD pod and run the following for each OSD, do you see a consistent value?
```
unset CEPH_ARGS
ceph daemon osd.0 config show  # change the id for each osd
```

**Comment by kbmanseau** [comment]
Thank you for the suggestion. It appears to report the same value as seen using the "tell" command from the toolbox. 
Output from inside the osd.0 pod itself:
`[root@k8s-master01 /]# ceph daemon osd.0 config show | grep osd_memory_target
    "osd_memory_target": "432633826508",`

**Comment by travisn** [comment]
Ok, that's consistent with a default of basically using unbounded memory. To confirm, setting the limits solves your memory issues, and this issue is just about if the defaults make sense? Rook doesn't set the default memory limits since every environment is so different.

**Comment by kbmanseau** [comment]
I will be implementing the limits later today after some testing. The issue was opened because the toolbox reports a different value for "osd_memory_target" than the OSDs. Shouldn't the [default](https://docs.ceph.com/en/latest/rados/configuration/bluestore-config-ref/#confval-osd_memory_target) of 4GB be set in the OSDs as well?

I ask because the default setting gives each OSD the allocation of the entire node (~400GB) resulting in massive memory over subscription (4OSDs x400GB memory allocation per node = 1600GB which is way more than each node has). For our workload the cache grows until the nodes run out of memory and then the OSDs crash, resulting in the process starting again

**Comment by travisn** [comment]
It is recommended to set the memory requests/limits on the OSDs, it's just not set by default. 

The toolbox showing the incorrect setting sounds like a bug in ceph.

**Comment by BlaineEXE** [comment]
If you supply resource requests/limits to the OSD pods via the CephCluster CR, then the OSDs should pick up on that by reading the env var on the pod that tells the request/limit.

Added to Ceph in PR: https://github.com/ceph/ceph/pull/29511
Code in Ceph: https://github.com/ceph/ceph/blob/16e152e161ffb6c57e4ed7d1b8a05e180432266b/src/common/config.cc#L496-L536

**Comment by kbmanseau** [comment]
Thank you for the information @BlaineEXE. I can confirm that after adding limits the issue is resolved. This was expected as you mentioned in the docs, however, my question is should there be default behavior to prevent this? Perhaps a default POD_MEMORY_LIMIT value to prevent issues like this from re-occurring? 

If not I think this can be closed

**Comment by kbmanseau** [comment]
Actually why is there still a discrepancy here?

- `ceph config get osd.1 osd_memory_target: 4294967296`
- `ceph tell osd.1 config show | grep osd_memory_target: "osd_memory_target": "12800000000", "osd_memory_target_cgroup_limit_ratio": "0.800000",`

The pod limits are set to 16G so 12.8G soft target with 80% ratio makes sense but why does "ceph config get" still report a different value

**Comment by BlaineEXE** [comment]
That's a good question. To be sure, it can be confusing without closely looking at Ceph's docs: https://docs.ceph.com/en/latest/rados/configuration/ceph-conf/#commands

From the docs:
> `ceph config get <who>` will dump the configuration for a specific daemon or client (e.g., mds.a), as stored in the monitors’ configuration database.

What "stored in the configuration database" means is that it will essentially show the user-specified config, but it won't necessarily show the running config.

The docs also say:
> ceph config show <who> will show the reported running configuration for a running daemon. These settings may differ from those stored by the monitors if there are also local configuration files in use or options have been overridden on the command line or at run time. The source of the option values is reported as part of the output.

This is why `ceph config show` is "right" and `ceph config get` is "wrong" in this case.

**Comment by kbmanseau** [comment]
Thank you @BlaineEXE for the additional info. I am not sure what can be done to avoid this confusion by others in the future but thanks to everyone's help we resolved our issue.

For others that may be interested we did some qualification of the affects setting limits on these pods has on the cluster performance. The specs of our cluster:

- 3xnodes with 524GB of memory each
- Each node has 4 OSDs for a total of 12
- The tests were run using a modified version of [arthur0/fio-tools](https://github.com/arthur0/fio-tools)
- Test runs all used 9g size

The graphs depict the performance when the OSD limits are set to 4g, 8g, 16g and no limit
![image](https://user-images.githubusercontent.com/22695969/146394811-0f8e160f-67f9-4536-9eca-b658e4f7c24c.png)
![image](https://user-images.githubusercontent.com/22695969/146394851-86be8ef9-8214-4c93-8aa8-91033b94ff19.png)
